### PR TITLE
undervolt: init at 0.2.8

### PR DESCRIPTION
**Pull request by jakefwoods**
###### Motivation for this change

Undervolt is a tool that lets you configure the voltage supplied to Intel CPUs on Linux. This is particularly useful for reducing the thermal footprint of a laptop.

This change _almost_ fits CONTRIBUTING.md but I'm not sure who the maintainer should be. Is that something I should add myself to?

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/os-specific/linux/undervolt/default.nix
Comment (samueldr) [inline pkgs/os-specific/linux/undervolt/default.nix]: While it builds on ARM, I'm pretty sure it won't work as it's for intel hardware. Furthermore, I would be surprised if it did anything for 32-bit processors. Might be better to only target x86_64

```
platforms = [ "x86_64-linux" ];
```

File: nixos/modules/services/hardware/undervolt.nix
Comment (samueldr) [inline nixos/modules/services/hardware/undervolt.nix]: ~~(In the following options) it might be better to drop the `Offset` suffix from the option names, pretty redundant and they're all offsets, except for `enable` and `package`.~~ Disregard that as there are more than voltage offsets.

It may be worthwhile to look into supporting other options like `--temp`

**Review by samueldr — COMMENTED** [review]
As I said, looks fine. Not approving because of my nitpicks, but not outright requesting changes in case another reviewer feels I'm overly zealous.

File: nixos/modules/services/hardware/undervolt.nix
Comment (Vodurden) [reply nixos/modules/services/hardware/undervolt.nix]: Good idea! I've added temp support and tested that it works on my NixOS machine

**Comment by eadwu** [comment]
Would be better to bump the version to 0.2.8. At least for my dear xps (cough https://github.com/georgewhewell/undervolt/issues/25).

**Comment by Vodurden** [comment]
Done! I tested 0.2.8 on my machine and it seems to be working fine. Let me know if there are any other issues \:)

**Comment by Vodurden** [comment]
@samueldr, @eadwu: Just checking in to see if there's anything else I need to do on this PR to get it merged. Please let me know if there are any other blockers and I'll try to resolve them as soon as possible. 

Thanks in advance \:)

**Review by samueldr — CHANGES_REQUESTED** [review]
Looks good, can't *actually* test it, but seeing there are two users already is good.

Just re-reading the code and saw some stuff, do comment/resolve as needed and I think then it'll be good to merge.

**Comment by Vodurden** [comment]
@samueldr: Thanks for the review! I've implemented your recommendations.

I also noticed that the logging in `journalctl -u undervolt` was _very_ verbose as I was always setting the `-v` flag so I've moved that behind an option. I've tested this new setting and everything seems fine.

Finally I've rebased this against master so hopefully there shouldn't be any conflicts if we feel it's ready to merge.

Thanks again and please let me know if there's anything else preventing this from being merged \:)

**Review by samueldr — APPROVED** [review]

**Comment by Vodurden** [comment]
@samueldr: Thanks for approving the changes! 

I'm just wondering if there's anything else that needs to be done to merge this branch? I don't have permission to merge so I'm not sure if I should just wait for someone else to click the merge button or if I need to make any changes/reach out to anyone to get it sorted out.

**Comment by Atemu** [comment]
Hey @Vodurden, is there a specific reason the undervolt needs to be applied every 2min?

The service activation is spamming my journal.

**Comment by Vodurden** [comment]
Hi @Atemu, I don't think there is any reason it needs to be 2 minutes. I copied the duration from the README of undervolt: https://github.com/georgewhewell/undervolt 

Re-reading the README it looks like they actually use a 30 second interval. I'm not sure why it needs to be so short but at the time I wrote this PR I didn't dig into the reason why. I wouldn't be surprised if just setting the value on boot works fine but you'd need to test it to be sure \:)

**Comment by Atemu** [comment]
Thanks! 

I found at least one while hacking around on it, the voltages are reset after resume.

**Comment by eadwu** [comment]
Yeah that's the reason why there's a timer, I use an alternate version.

```
    systemd.services.undervolt = {
      after = [ "suspend.target" "hibernate.target" "hybrid-sleep.target" ];
      wantedBy = [ "suspend.target" "hibernate.target" "hybrid-sleep.target" "multi-user.target" ];

      description = "Intel Undervolt";
      serviceConfig.ExecStart = undervoltCmd;
    };

    services.udev.extraRules = ''
      # Apply an undervolt for Intel CPUs
      ACTION=="change", SUBSYSTEM=="power_supply", ATTR{type}=="Mains", ATTR{online}=="0", RUN+="${undervoltCmd}"
      ACTION=="change", SUBSYSTEM=="power_supply", ATTR{type}=="Mains", ATTR{online}=="1", RUN+="${undervoltCmd}"
    '';
```

**Comment by Atemu** [comment]
Thanks! I experimented around a bit with this and found that the first declaration can be condensed down into 

```Nix
  systemd.services.undervolt.wantedBy = [ "post-resume.target" "multi-user.target" ];
  systemd.services.undervolt.after = [ "post-resume.target" ];
```

What exactly does the second one do? I'm not too familiar with udev but it looks like it's supposed to deal with power supply changes?  
I couldn't get the undervolt to go away when (dis-)connecting the charger though.

**Comment by eadwu** [comment]
It may be a problem with certain CPUs or computers, anyway if you need more help create a new issue instead of commenting here.

**Pull request merged**
